### DISCUSSION TITLE
Function parameter analyzers

### DISCUSSION THREAD

**Discussion by bcollamore** [philips-software/roslyn-analyzers#61] [Ideas]
See https://github.com/philips-software/roslyn-analyzers/issues/58 for conversation start.

**Comment by bcollamore** [comment]
* "In the case of Ints or String parameters, if only 3 values are ever used we can suggest the use of an enum."  This idea exposes a limitation of Roslyn: it only works within a project.  Therefore, this idea would only benefit for internal or private methods.  But it is still interesting and do-able. 

* "In the case of methods that is overloaded with types that don't share a hierarchy we can suggest to create an interface."
This is do-able, and a natural application of Roslyn.

* "We can create a warning for types that have null fields."
Can you elaborate?

**Reply by kryptt** [reply]
For the third point, 

Its about an alternative approach to https://www.affirmaconsulting.com/null-values-in-c/ in that we look for opportunities to separate a class with optional / null fields into two separate classes that inherit from a base / share an interface.. basically the example used in the article that  replaces a big class with a couple of parameters that could be null with two separate sibling classes that each express one of the possiblities

**Comment by kryptt** [comment]
Here is a long workshop on all the ways to find the different counts:

 https://github.com/cohomolo-gy/Isomorphic-Reasoning

We probably don't need this since parametricity is virtually non-existent in C#
